File: app/src/main/java/com/stupidtree/hichat/ui/widgets/PhotoDetailActivity.java
package com.stupidtree.hichat.ui.widgets;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.bm.library.PhotoView;
import com.bumptech.glide.Glide;
import com.stupidtree.hichat.R;
import com.stupidtree.hichat.ui.base.BaseActivity;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import butterknife.BindView;


public class PhotoDetailActivity extends BaseActivity<ViewModel> {

    @BindView(R.id.label)
    TextView label;
    @BindView(R.id.pager)
    ViewPager pager;
    int initIndex;


    List<String> urls;
    @Override
    protected int getLayoutId() {
        return R.layout.activity_photo_detail;
    }

    @Override
    protected Class<ViewModel> getViewModelClass() {
        return null;
    }

    @SuppressLint("SetTextI18n")
    @Override
    protected void initViews() {
        urls = new ArrayList<>();
        String[] data = getIntent().getStringArrayExtra("urls");
        initIndex = getIntent().getIntExtra("init_index", 0);
        if (data != null) {
            urls.addAll(Arrays.asList(data));
            label.setText((initIndex + 1) + "/" + urls.size());
            initPager();
        }
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setWindowParams(true, false, true);

    }

    void initPager() {
        pager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @SuppressLint("SetTextI18n")
            @Override
            public void onPageSelected(int position) {
                label.setText((position + 1) + "/" + urls.size());
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
        pager.setAdapter(new PagerAdapter() {
            @Override
            public int getCount() {
                return urls.size();
            }

            @Override
            public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
                return view == object;
            }

            @Override
            public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
                container.removeView((View) object);
            }

            @NotNull
            @Override
            public Object instantiateItem(@NotNull ViewGroup container, int position) {
                PhotoView v = new PhotoView(getThis());
                v.setScaleType(ImageView.ScaleType.FIT_CENTER);
                v.setTransitionName("image");
                Glide.with(getThis()).load(urls.get(position)).timeout(10000).into(v);
                v.setAdjustViewBounds(false);
                v.enable();
                container.addView(v);
                v.setOnClickListener(view -> finish());
//                v.setOnLongClickListener(new View.OnLongClickListener() {
//                    @Override
//                    public boolean onLongClick(View v) {
//                        new AlertDialog.Builder(getThis()).setItems(new String[]{getString(R.string.download_image)}, new DialogInterface.OnClickListener() {
//                            @Override
//                            public void onClick(DialogInterface dialog, int which) {
//                                // Toast.makeText(from,imageViewToUrl.get(view)+"",Toast.LENGTH_SHORT).show();
//                                ActivityUtils.DownloadImage(getThis(), urls.get(which), new ActivityUtils.OnDownloadDoneListener() {
//                                    @Override
//                                    public void onDone() {
//                                        try {
//                                            Toast.makeText(getThis(), R.string.save_done, Toast.LENGTH_SHORT).show();
//
//                                        } catch (Exception e) {
//
//                                        }
//                                    }
//                                });
//                            }
//                        }).create().show();
//                        return true;
//                    }
//                });
                return v;
            }
        });
        pager.setCurrentItem(initIndex);
    }


}

File: app/src/main/java/com/stupidtree/cloudliter/ui/widgets/PhotoDetailActivity.kt
package com.stupidtree.cloudliter.ui.widgets

import android.annotation.SuppressLint
import android.app.Application
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.core.view.drawToBitmap
import androidx.lifecycle.AndroidViewModel
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.bm.library.PhotoView
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.stupidtree.cloudliter.R
import com.stupidtree.cloudliter.databinding.ActivityPhotoDetailBinding
import com.stupidtree.cloudliter.ui.imagedetect.ImageDetectFragment
import com.stupidtree.style.base.BaseActivity

class PhotoDetailActivity : BaseActivity<PhotoDetailActivity.PhotoViewModel, ActivityPhotoDetailBinding>() {

    var initIndex = 0
    private val ids: MutableList<String> = mutableListOf()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding.toolbar.title = ""
        setToolbarActionBack(binding.toolbar)
    }

    @SuppressLint("SetTextI18n")
    override fun initViews() {
        val data = intent.getStringArrayExtra("ids")
        initIndex = intent.getIntExtra("init_index", 0)

        if (data?.isNotEmpty() == true) {
            ids.addAll(listOf(*data))
            binding.label.text = (initIndex + 1).toString() + "/" + ids.size
            initPager()
        }
        supportFragmentManager.beginTransaction()
                .replace(R.id.detect_layout, ImageDetectFragment.newInstance(ids[initIndex], local = false, refreshOnStart = true), "detect").commit()
    }

    private fun initPager() {
        binding.pager.addOnPageChangeListener(object : OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

            @SuppressLint("SetTextI18n")
            override fun onPageSelected(position: Int) {
                binding.label.text = (position + 1).toString() + "/" + ids.size
                (supportFragmentManager.findFragmentByTag("detect")
                        as ImageDetectFragment?)?.preLoadImage(ids[position])


            }

            override fun onPageScrollStateChanged(state: Int) {}
        })
        binding.pager.adapter = object : PagerAdapter() {
            override fun getCount(): Int {
                return ids.size
            }

            override fun isViewFromObject(view: View, `object`: Any): Boolean {
                return view === `object`
            }

            override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
                container.removeView(`object` as View)
            }

            override fun instantiateItem(container: ViewGroup, position: Int): Any {
                val v = PhotoView(getThis())
                v.scaleType = ImageView.ScaleType.FIT_CENTER
                v.transitionName = "image"
                v.contentDescription = getString(R.string.photo_description, position + 1, ids.size)
                com.stupidtree.cloudliter.utils.ImageUtils.loadCloudImageInto(getThis(), ids[position], v)
                v.adjustViewBounds = false
                v.contentDescription = getString(R.string.image)
                v.enable()
                container.addView(v)
                return v
            }
        }
        binding.bar.setOnClickListener {
            val behavior = BottomSheetBehavior.from(binding.bottomLayout)
            if (behavior.state == BottomSheetBehavior.STATE_EXPANDED) {
                behavior.state = BottomSheetBehavior.STATE_COLLAPSED
                binding.bar.contentDescription = getString(R.string.detect_expand)
            } else {
                behavior.state = BottomSheetBehavior.STATE_EXPANDED
                binding.bar.contentDescription = getString(R.string.detect_collapse)
            }
            behavior.addBottomSheetCallback(object : BottomSheetBehavior.BottomSheetCallback() {
                override fun onStateChanged(bottomSheet: View, newState: Int) {
                    if (newState == BottomSheetBehavior.STATE_EXPANDED) {
                        (supportFragmentManager.findFragmentByTag("detect")
                                as ImageDetectFragment?)?.refresh(false, ids[binding.pager.currentItem])
                    }
                }

                override fun onSlide(bottomSheet: View, slideOffset: Float) {

                }

            })
        }
        binding.pager.currentItem = initIndex
    }

    class PhotoViewModel(application: Application) : AndroidViewModel(application)

    override fun initViewBinding(): ActivityPhotoDetailBinding {
        return ActivityPhotoDetailBinding.inflate(layoutInflater)
    }

    override fun getViewModelClass(): Class<PhotoViewModel> {
        return PhotoViewModel::class.java
    }
}